corehq/apps/sso/forms.py

```diff
@@ -218,7 +218,30 @@ def service_provider_fields(self):
         return shown_fields
 
 
-class EditIdentityProviderAdminForm(forms.Form):
+class _UpdateIdentityProviderFormMixin:
```

Comment:
A couple thoughts on this mixin.

First, I'm not sure this logic is worth it. From looking at the identity provider model, it seems unlikely that any attribute is likely to change very frequently -- we could just clear the cache on any modification and ignore determining what changed.

Second, what advantage does a mixin provide over composition? Without going into a huge detour, here are my points against mixins:
- suffer from the multiple inheritance diamond problem
- are harder to read. It isn't obvious what code lives where, and then you're jumping back and forth between mixins to understand how your class is getting mutated
- harder to test. Mixins need to hijack another class to be test-able.
- They're somewhat language-exclusive. Many modern languages don't support the concept.

Instead, this could just be a separate function that takes an identity provider, an admin_user, and possibly a form.

Reply:
this logic is absolutely worth it, as we have to manually populate `update_fields`—this is not automatically populated. so I decided to create a **private** mixin for this use case to avoid duplication of code and make sure that the save action is consistent and does this very important task so that the signal picks it up.

Reply:
The need to populate `update_fields` was the logic I was referring to -- it seems like overkill. If this was a high-traffic area that had an identity provider changing thousands of times per second, then I'd understand the need to limit updates. But Identity Providers will likely change what? Four times a year? I'm framing extra code as a liability -- someone needs to understand and maintain it. In the rare event that an IdP changes, it seems simpler just to always clear the cache.

The Mixin thing might just be my issue. I don't like that I have to hop around figuring out what attributes are coming from where, and I don't like that a mixin itself looks incomplete because it relies on the actual classes to be in the correct state.  When I see a definition like `class A(B, C)`, I read it as saying A IS a B and a C. Mixins break that. An `EditIdentityProviderAdminForm` is not a `_UpdateIdentityProviderFormMixin` -- it just uses its functionality. If there exists enough common SSO functionality between `EditIdentityProviderAdminForm` and `SSOEnterpriseSettingsForm`, then it starts to make sense to create something like a base `SSOForm`, and then we're back to single inheritance. If not, both classes could use composition -- create their own `update_identity_provider` methods that simply call the same private helper function. But again...maybe this is my own issue with Python and I have to get over it.

Reply:
ok, I can agree with the argument to clear caches on save. :+1: